include/SFML/Window/Event.hpp

```diff
@@ -34,196 +34,298 @@
 #include <SFML/Window/Mouse.hpp>
 #include <SFML/Window/Sensor.hpp>
 
+#include <type_traits>
+#include <variant>
+
+#include <cassert>
 
-// NOLINTBEGIN(cppcoreguidelines-pro-type-member-init)
 
 namespace sf
 {
 ////////////////////////////////////////////////////////////
 /// \brief Defines a system event and its parameters
 ///
 ////////////////////////////////////////////////////////////
-struct Event
+class Event
 {
+public:
+    struct Empty
+    {
+    };
+
+    struct Closed
+    {
+    };
+
+    struct LostFocus
+    {
+    };
+
+    struct GainedFocus
+    {
+    };
+
     ////////////////////////////////////////////////////////////
     /// \brief Size events parameters (Resized)
     ///
     ////////////////////////////////////////////////////////////
-    struct SizeEvent
+    struct Resized
     {
-        unsigned int width;  //!< New width, in pixels
-        unsigned int height; //!< New height, in pixels
+        unsigned int width{};  //!< New width, in pixels
+        unsigned int height{}; //!< New height, in pixels
```

Comment:
I've wanted to do that for a while. I forget why but last time I tried that a long time ago it got messy. I'm open to trying that again since it would be really nice to remove most of these `width`/`height` and `x`/`y` pairs.

Reply:
I'd definitely go with `Vector2` here, even if it's just to have a simple way to pass it along other functions expecting a `Vector2` (same for other similar event data).

Reply:
I like this idea and want to play with it again but this PR is already pretty big so I'd rather not get into this as to ensure the scope of the PR doesn't expand too much more. This will be a great PR for later when we can make this change in isolation.

Reply:
https://github.com/SFML/SFML/compare/event_api...resized_event

This change works very well in practice. I'm willing to add this to this PR but won't yet for the sake of keeping scope small. This change will certainly happen at some point though.